src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -16,8 +16,96 @@ extern "C" {
 
 }
 
+#include <qurt.h>
+
 #include "../HalideRuntime.h"
 
+const int stack_alignment = 128;
+const int stack_size = 1024 * 1024;
+
+typedef int (*pipeline_argv_t)(void **);
+
+// We can't control the stack size on the thread which receives our
+// FastRPC calls. To work around this, we make our own thread, and
+// forward the calls to that thread.
+class PipelineContext {
```

Comment:
Here's a sketch:

`class PipelineContext {
    void *stack;
    qurt_thread_t thread;
    qurt_cond_t work_state_change;
    qurt_cond_t work_completed;
    qurt_mutex_t mutex;

    // Shared state
    bool still_running;
    pipeline_argv_t function;
    void **args;
    int result;

    void thread_main() {
        qurt_mutex_lock(&mutex);
        while (still_running) { // Favor quiting over executing pending item.
	    while (still_running && function == nullptr) {
	        qurt_cond_wait(&work_state_change, &mutex);
            }
	    if (funtion != nullptr) {
	        result = function(args);
		function = nullptr;
                qurt_cond_signal(&work_completed);
            }
        }
        qurt_mutex_unlock(&mutex);
    }

    static void redirect_main(void *data) {
        static_cast<PipelineContext *>(data)->thread_main();
    }

 +public:
    PipelineContext(int stack_alignment, int stack_size)
        : stack(NULL), still_running(true), function(NULL), args(NULL) {
        qurt_mutex_init(&work_mutex);
        qurt_mutex_init(&wakeup_master_mutex);
        qurt_cond_init(&work_state_change);
        qurt_cond_init(&work_completed);

        // Allocate the stack for this thread.
        stack = memalign(stack_alignment, stack_size);

        qurt_thread_attr_t thread_attr;
        qurt_thread_attr_init(&thread_attr);
        qurt_thread_attr_set_stack_addr(&thread_attr, stack);
        qurt_thread_attr_set_stack_size(&thread_attr, stack_size);
        qurt_thread_attr_set_priority(&thread_attr, 100);
        qurt_thread_create(&thread, &thread_attr, redirect_main, this);
    }

    ~PipelineContext() {
        still_running = false;
	qurt_cond_signal(&work_state_change);

        // Status is probably garbage. If we want it to be something, we'll need a thread exit in thread_main.
        int status;
        qurt_thread_join(thread, &status);

        free(stack);
    }

    int run(pipeline_argv_t function, void **args) {
        int local_result;

        // get a lock and set up work for the worker.
        qurt_mutex_lock(&work_mutex);
        this->function = function;
        this->args = args;

        // send a signal to the worker.
	qurt_cond_signal(&work_state_change);
	while (function != nullptr) {
	    qurt_cond_wait(&work_completed, &mutex);
        }
	local_result = result;
        qurt_mutex_unlock(&mutex);

        return local_result;
    }
};`